src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,212 @@
+  }
+
+  RAY_CHECK(task_spec.IsActorTask());
+  const auto &actor_id = task_spec.ActorId();
+
+  const auto task_id = task_spec.TaskId();
+  const auto num_returns = task_spec.NumReturns();
+
+  auto request = std::unique_ptr<rpc::PushTaskRequest>(new rpc::PushTaskRequest);
+  request->set_task_id(task_spec.TaskId().Binary());
+  request->mutable_task_spec()->CopyFrom(task_spec.GetMessage());
+
+  std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+  auto iter = actor_states_.find(actor_id);
+  if (iter == actor_states_.end() ||
+      iter->second.state_ == ActorTableData::RECONSTRUCTING) {
+    // Actor is not yet created, or is being reconstructed, cache the request
+    // and submit after actor is alive.
+    // TODO(zhijunfu): it might be possible for a user to specify an invalid
+    // actor handle (e.g. from unpickling), in that case it might be desirable
+    // to have a timeout to mark it as invalid if it doesn't show up in the
+    // specified time.
+    auto pending_request = std::unique_ptr<PendingTaskRequest>(
+        new PendingTaskRequest(task_id, num_returns, std::move(request)));
+    pending_requests_[actor_id].emplace_back(std::move(pending_request));
+    return Status::OK();
+  } else if (iter->second.state_ == ActorTableData::ALIVE) {
+    // Actor is alive, submit the request.
+    if (rpc_clients_.count(actor_id) == 0) {
+      // If rpc client is not available, then create it.
+      HandleActorAlive(actor_id, iter->second.location_.first,
+                       iter->second.location_.second);
+    }
+
+    // Submit request.
+    auto &client = rpc_clients_[actor_id];
+    return PushTask(*client, *request, task_id, num_returns);
+  } else {
+    // Actor is dead, treat the task as failure.
+    RAY_CHECK(iter->second.state_ == ActorTableData::DEAD);
+    TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+    return Status::IOError("actor is dead or being reconstructed");
+  }
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubscribeActorUpdates() {
+  // Register a callback to handle actor notifications.
+  auto actor_notification_callback = [this](const ActorID &actor_id,
+                                            const ActorTableData &actor_data) {
+    std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+    actor_states_.erase(actor_id);
+    actor_states_.emplace(
+        actor_id,
+        ActorStateData(actor_data.state(), actor_data.ip_address(), actor_data.port()));
+
+    if (actor_data.state() == ActorTableData::ALIVE) {
+      // Check if this actor is the one that we're interested, if we already have
+      // a connection to the actor, or have pending requests for it, we should
+      // create a new connection.
+      if (rpc_clients_.count(actor_id) > 0 || pending_requests_.count(actor_id) > 0) {
+        HandleActorAlive(actor_id, actor_data.ip_address(), actor_data.port());
+      }
+    }
```

Comment:
we should clean up the client when an actor is dead or being reconstructed.

Reply:
please kindly refer to my reply in the next comment.

Reply:
after looking it more closely, yes this can be cleaned up